driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -333,11 +334,28 @@ public Session connect(String keyspace) {
      * {@link #closeAsync()}), or as a result of an error while initializing the
      * Cluster.
      */
-    public ListenableFuture<Session> connectAsync(String keyspace) {
+    public ListenableFuture<Session> connectAsync(final String keyspace) {
         checkNotClosed(manager);
         init();
-        Session session = manager.newSession(keyspace);
-        return session.initAsync();
+        final Session session = manager.newSession();
+        ListenableFuture<Session> sessionInitialized = session.initAsync();
```

Comment:
I like this because it now becomes fairly simple that if setting a keyspace fails, the future fails and the session is closed.  This could lead to some false positives, i.e. if you hit a client timeout here the session can't be established...but that seems fair..